File: app/src/main/java/com/wolcano/musicplayer/music/ui/adapter/PlaylistAdapter.java
package com.wolcano.musicplayer.music.ui.adapter;

import android.content.Context;
import androidx.annotation.NonNull;
import androidx.appcompat.widget.PopupMenu;
import androidx.core.content.ContextCompat;
import androidx.recyclerview.widget.RecyclerView;

import android.view.ContextThemeWrapper;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.ImageView;
import android.widget.TextView;
import android.widget.Toast;
import com.afollestad.materialdialogs.MaterialDialog;
import com.wolcano.musicplayer.music.R;
import com.wolcano.musicplayer.music.mvp.models.Playlist;
import com.wolcano.musicplayer.music.utils.SongUtils;
import com.wolcano.musicplayer.music.utils.Utils;

import java.util.List;

public class PlaylistAdapter  extends RecyclerView.Adapter<RecyclerView.ViewHolder> {

    private List<Playlist> arraylist;
    private Context context;

    public PlaylistAdapter(Context context, List<Playlist> arraylist){
        this.context = context;
        this.arraylist = arraylist;
    }

    @NonNull
    @Override
    public RecyclerView.ViewHolder onCreateViewHolder(@NonNull ViewGroup parent, int viewType) {
        RecyclerView.ViewHolder viewHolder;
        View v;
        v = LayoutInflater.from(parent.getContext()).inflate(R.layout.item_playlist, parent, false);
        viewHolder = new PlaylistAdapter.ViewHolder(v);
        return viewHolder;
    }

    @Override
    public void onBindViewHolder(@NonNull RecyclerView.ViewHolder holder, int position) {
        PlaylistAdapter.ViewHolder viewHolder = (PlaylistAdapter.ViewHolder) holder;

        Playlist playlist = arraylist.get(position);

        viewHolder.line2.setText(Utils.createStr(context, R.plurals.Nsongs, playlist.getSongCount()));
        viewHolder.line1.setText(playlist.getName());
        viewHolder.albumArt.setColorFilter(ContextCompat.getColor(context,R.color.grey0));
        setOnPopupMenuListener(viewHolder, position);


    }
    private void setOnPopupMenuListener(PlaylistAdapter.ViewHolder holder, final int position) {
        holder.more.setOnClickListener(v -> {
            try {
                ContextThemeWrapper contextThemeWrapper = new ContextThemeWrapper(v.getContext(), R.style.PopupMenuToolbar);

                PopupMenu popup = new PopupMenu(contextThemeWrapper, v);
                popup.getMenuInflater().inflate(R.menu.popup_menu_playlist, popup.getMenu());
                popup.show();
                popup.setOnMenuItemClickListener(item -> {
                    switch (item.getItemId()) {
                        case R.id.rename:
                            new MaterialDialog.Builder(context)
                                    .title(R.string.rename)
                                    .positiveText(R.string.change)
                                    .negativeText(R.string.cancel)
                                    .positiveColor(Utils.getAccentColor(context))
                                    .negativeColor(Utils.getAccentColor(context))
                                    .input(null, arraylist.get(position).getName(), false, new MaterialDialog.InputCallback() {
                                        @Override
                                        public void onInput(@NonNull MaterialDialog dialog, CharSequence input) {
                                            SongUtils.renamePlaylist(context,arraylist.get(position).getId(), input.toString());
                                            arraylist.get(position).setName(input.toString());
                                            holder.line1.setText(input.toString());
                                            Toast.makeText(context, R.string.rename_playlist_success, Toast.LENGTH_SHORT).show();
                                        }
                                    })
                                    .show();
                            break;
                        case R.id.delete:
                            new MaterialDialog.Builder(context)
                                    .title(arraylist.get(position).getName())
                                    .content(R.string.delete_playlist)
                                    .positiveText(R.string.delete)
                                    .negativeText(R.string.cancel)
                                    .positiveColor(Utils.getAccentColor(context))
                                    .negativeColor(Utils.getAccentColor(context))
                                    .onPositive((dialog, which) -> {
                                        SongUtils.deletePlaylists(context,arraylist.get(position).getId());
                                        arraylist.remove(holder.getAdapterPosition());
                                        notifyItemRemoved(holder.getAdapterPosition());
                                        notifyItemRangeChanged(holder.getAdapterPosition(), getItemCount() - holder.getAdapterPosition());

                                        Toast.makeText(context, R.string.delete_playlist_success, Toast.LENGTH_SHORT).show();
                                    })
                                    .onNegative((dialog, which) -> dialog.dismiss())
                                    .show();
                            break;
                    }
                   return true;
                });
            } catch (Exception e) {
                e.printStackTrace();
            }
        });
    }
    @Override
    public int getItemCount() {
        return (null != arraylist ? arraylist.size() : 0);
    }


    public class ViewHolder extends RecyclerView.ViewHolder implements View.OnClickListener {
        TextView line1;
        TextView line2;
        ImageView albumArt, more;
        public ViewHolder(View view) {
            super(view);
            this.line1 =  view.findViewById(R.id.line1);
            this.line2 =  view.findViewById(R.id.line2);
            this.albumArt =  view.findViewById(R.id.playlistImg);
            this.more = view.findViewById(R.id.more);
            view.setOnClickListener(this);
        }

        @Override
        public void onClick(View v) {
            Playlist playlist = arraylist.get(getAdapterPosition());
                long playlistID = playlist.getId();
                String playlistName = playlist.getName();
                Utils.navigateToPlaylist(context, playlistID,
                        playlistName);

        }
    }
}

File: app/src/main/java/com/wolcano/musicplayer/music/ui/adapter/PlaylistAdapter.kt
package com.wolcano.musicplayer.music.ui.adapter

import android.content.Context
import android.view.*
import android.widget.Toast
import androidx.appcompat.widget.PopupMenu
import androidx.core.content.ContextCompat
import androidx.databinding.DataBindingUtil
import androidx.recyclerview.widget.RecyclerView
import com.afollestad.materialdialogs.MaterialDialog
import com.afollestad.materialdialogs.input.input
import com.wolcano.musicplayer.music.R
import com.wolcano.musicplayer.music.databinding.ItemPlaylistBinding
import com.wolcano.musicplayer.music.mvp.models.Playlist
import com.wolcano.musicplayer.music.utils.SongUtils
import com.wolcano.musicplayer.music.utils.Utils.createStr
import com.wolcano.musicplayer.music.utils.Utils.navigateToPlaylist

class PlaylistAdapter(val context: Context, val playlistList: MutableList<Playlist>) :
    RecyclerView.Adapter<PlaylistAdapter.ViewHolder>() {

    override fun onCreateViewHolder(parent: ViewGroup, viewType: Int): ViewHolder {
        val binding: ItemPlaylistBinding = DataBindingUtil.inflate(
            LayoutInflater.from(parent.context),
            R.layout.item_playlist,
            parent,
            false
        )
        return ViewHolder(binding)
    }

    override fun onBindViewHolder(holder: ViewHolder, position: Int) {
        holder.binding.playlist = playlistList[position]
        holder.binding.executePendingBindings()

        val playlist: Playlist? = holder.binding.playlist

        holder.binding.line1.text = playlist?.name
        holder.binding.line2.text = createStr(context, R.plurals.Nsongs, playlist!!.songCount)
        holder.binding.playlistImg.setColorFilter(ContextCompat.getColor(context, R.color.grey0))
        setOnPopupMenuListener(holder, position)
    }

    private fun setOnPopupMenuListener(holder: ViewHolder, position: Int) {
        holder.binding.more.setOnClickListener { v ->
            try {
                val contextThemeWrapper =
                    ContextThemeWrapper(v.context, R.style.PopupMenuToolbar)
                val popup =
                    PopupMenu(contextThemeWrapper, v)
                popup.menuInflater.inflate(R.menu.popup_menu_playlist, popup.menu)
                popup.show()
                popup.setOnMenuItemClickListener { item: MenuItem ->
                    when (item.itemId) {
                        R.id.rename ->
                            MaterialDialog(context).show {
                                title(R.string.rename)
                                positiveButton(R.string.change)
                                negativeButton(R.string.cancel)
                                input(playlistList[position].name) { _, input ->
                                    SongUtils.renamePlaylist(
                                        context,
                                        playlistList[position].id,
                                        input.toString()
                                    )
                                    playlistList[position].name = input.toString()
                                    holder.binding.line1.text = input.toString()
                                    Toast.makeText(
                                        context,
                                        R.string.rename_playlist_success,
                                        Toast.LENGTH_SHORT
                                    ).show()

                                }
                            }
                        R.id.delete ->
                            MaterialDialog(context).show {
                                title(text = playlistList[position].name)
                                message(R.string.delete_playlist)
                                positiveButton(R.string.delete) {
                                    SongUtils.deletePlaylists(
                                        context,
                                        playlistList[position].id
                                    )
                                    playlistList.removeAt(holder.adapterPosition)
                                    notifyItemRemoved(holder.adapterPosition)
                                    notifyItemRangeChanged(
                                        holder.adapterPosition,
                                        itemCount - holder.adapterPosition
                                    )
                                    Toast.makeText(
                                        context,
                                        R.string.delete_playlist_success,
                                        Toast.LENGTH_SHORT
                                    ).show()
                                }
                                negativeButton(R.string.cancel) {
                                    dismiss()
                                }
                            }
                    }
                    true
                }
            } catch (e: Exception) {
                e.printStackTrace()
            }
        }
    }

    override fun getItemCount(): Int {
        return playlistList.size
    }

    inner class ViewHolder(val binding: ItemPlaylistBinding) : RecyclerView.ViewHolder(
        binding.root
    ),
        View.OnClickListener {
        override fun onClick(v: View) {
            val playlist: Playlist = playlistList[adapterPosition]
            val playlistID = playlist.id
            val playlistName = playlist.name
            navigateToPlaylist(
                context, playlistID,
                playlistName
            )
        }

        init {
            binding.root.setOnClickListener(::onClick)
        }
    }

}